Regression: cross-module lookup of an associated type witnessed by a protocol-extension `typealias` fails in Swift 6.4

## Summary

A protocol declares `associatedtype Action`, and the witness is supplied by a
`typealias Action = …` in a **protocol extension** (parameterized by the
conforming type). A conformer in a second module satisfies the requirement
implicitly through that extension. Referencing `Conformer.Action` from a *third*
module fails to type-check under Swift 6.4:

```
error: 'Action' is not a member type of struct 'Mid.Foo'
```

The same sources compile cleanly under Swift 6.3.2. Same-module references
resolve on both compilers; only cross-module qualified lookup of the
extension-witnessed associated-type alias regressed.

This reproduces with `swiftc` directly (no build system, no SwiftPM, no Xcode),
so it is a compiler frontend regression.

## Reproducer

Three source files, three modules, no dependencies, pure language semantics.

`Core.swift`:
```swift
public protocol Feature {
    associatedtype State
    associatedtype Action
}

extension Feature {
    public typealias Action = GenericAction<Self, State>
}

public enum GenericAction<F: Feature, State> {
    case state(State)
}
```

`Mid.swift` (conformer — no explicit `Action`):
```swift
import Core

public struct Foo: Feature {
    public typealias State = Int
}
```

`App.swift` (reference from a third module):
```swift
import Core
import Mid

public func useFooAction(_ action: Foo.Action) {}
```

Build each as its own module:
```sh
SDK=$(xcrun --sdk macosx --show-sdk-path)
swiftc -sdk "$SDK" -emit-module -module-name Core Core.swift -emit-module-path Core.swiftmodule
swiftc -sdk "$SDK" -emit-module -module-name Mid  -I . Mid.swift  -emit-module-path Mid.swiftmodule
swiftc -sdk "$SDK" -emit-module -module-name App  -I . App.swift  -emit-module-path App.swiftmodule
```

(A SwiftPM `Package.swift` with three targets `Core` → `Mid` → `App` reproduces
the same failure via `swift build`.)

## Expected vs. actual

**Expected** (Swift 6.3.2): all three modules emit successfully.

**Actual** (Swift 6.4), the `App` module fails:
```
App.swift:3:40: error: 'Action' is not a member type of struct 'Mid.Foo'
1 | import Core
2 | import Mid
3 | public func useFooAction(_ action: Foo.Action) {}
  |                                        `- error: 'Action' is not a member type of struct 'Mid.Foo'

Mid.Foo:1:15: note: 'Foo' declared here
```

## The three-module split is required

Collapsing any two of the modules compiles on both 6.3.2 and 6.4:

- Core + Mid merged (conformer in the same module as the extension), `App` references it → compiles.
- Mid + App merged (reference in the same module as the conformer) → compiles.
- All in one module → compiles.

The failure needs all three of: the extension that supplies the witness, the
conformer, and the reference site each in a **separate** module, with the
reference site importing both the witness module (`Core`) and the conformer
module (`Mid`).

## Workaround

Declare the `typealias` explicitly on each conformer that is referenced
cross-module:
```swift
public struct Foo: Feature {
    public typealias State = Int
    public typealias Action = GenericAction<Self, State>   // workaround
}
```
This compiles on 6.4. It has to be repeated on every conformer, which defeats
the purpose of supplying the witness once in a shared extension.

## Not a regression (scope guard)

A superficially similar failure is *not* a regression — it fails identically on
6.3.2 and 6.4: if the reference site imports only `Mid` but not `Core` (the
module that declares the extension), lookup fails with `cannot use type alias
'Action' here; 'Core' was not imported`. That is expected behavior. The
regression reported here specifically requires the reference site to import the
witness module and still fail member-type lookup on the cross-module conformer.

## Environment

Regressed in:
```
Apple Swift version 6.4 (swiftlang-6.4.0.20.104 clang-2100.3.20.102)
```
Last known good:
```
Apple Swift version 6.3.2 (swiftlang-6.3.2.1.108 clang-2100.1.1.101)
```
Sources and build commands are byte-identical between the two runs; only the
compiler differs. (These ship in Xcode 27.0 Beta 1 `27A5194q` and Xcode 26.5
respectively, selected via `DEVELOPER_DIR`.) Host: Apple Silicon, macOS.

--- a/lib/Serialization/Deserialization.cpp
+++ b/lib/Serialization/Deserialization.cpp
@@ -9266,16 +9266,6 @@ void ModuleFile::finishNormalConformance(NormalProtocolConformance *conformance,
     } else {
       fatal(thirdOrError.takeError());
     }
-    if (isa_and_nonnull<TypeAliasDecl>(third) &&
-        third->getModuleContext() != getAssociatedModule() &&
-        !third->getDeclaredInterfaceType()->isEqual(second)) {
-      // Conservatively drop references to typealiases in other modules
-      // that may have changed. This may also drop references to typealiases
-      // that /haven't/ changed but just happen to have generics in them, but
-      // in practice having a declaration here isn't actually required by the
-      // rest of the compiler.
-      third = nullptr;
-    }
     typeWitnesses[first] = {second, third};
   }
   assert(rawIDIter <= rawIDs.end() && "read too much");

--- a/test/AssociatedTypeInference/cross_module_extension_typealias_witness.swift
+++ b/test/AssociatedTypeInference/cross_module_extension_typealias_witness.swift
@@ -1,0 +1,57 @@
+// RUN: %empty-directory(%t)
+// RUN: split-file %s %t
+
+// Build three separate modules: Core (declares the protocol and the
+// protocol-extension type alias that witnesses the associated type), Mid (the
+// conformer, which does *not* declare the associated type explicitly), and App
+// (which references the conformer's associated type and imports both modules).
+
+// RUN: %target-swift-frontend -emit-module -module-name Core %t/Core.swift -emit-module-path %t/Core.swiftmodule
+// RUN: %target-swift-frontend -emit-module -module-name Mid %t/Mid.swift -emit-module-path %t/Mid.swiftmodule -I %t
+// RUN: %target-swift-frontend -typecheck -verify %t/App.swift -I %t
+
+// https://github.com/swiftlang/swift/issues/89909
+//
+// Cross-module qualified lookup of an associated type whose witness is supplied
+// by a type alias in a protocol extension used to fail when the protocol
+// extension, the conformer, and the reference site each lived in a separate
+// module. Deserializing the conformer's module used to drop the witness type
+// alias declaration (because the generic type alias lives in yet another
+// module), which left qualified lookup with no declaration to resolve, so
+// 'Foo.Action' failed with "'Action' is not a member type of struct 'Foo'".
+
+//--- Core.swift
+
+public protocol Feature {
+    associatedtype State
+    associatedtype Action
+}
+
+extension Feature {
+    public typealias Action = GenericAction<Self, State>
+}
+
+public enum GenericAction<F: Feature, State> {
+    case state(State)
+}
+
+//--- Mid.swift
+
+import Core
+
+public struct Foo: Feature {
+    public typealias State = Int
+}
+
+//--- App.swift
+
+import Core
+import Mid
+
+// The associated type 'Action' resolves to the extension-supplied witness
+// 'GenericAction<Foo, Int>'.
+public func useFooAction(_ action: Foo.Action) {}
+
+func checkWitnessType() {
+    let _: Foo.Action = GenericAction<Foo, Int>.state(0)
+}

